backends/gstreamer/media_stream_source.rs

```diff
@@ -50,21 +59,45 @@ mod imp {
         // Called once at the very beginning of instantiation of each instance and
         // creates the data structure that contains all our state
         fn new_with_class(klass: &subclass::simple::ClassStruct<Self>) -> Self {
-            let proxysrc = gst::ElementFactory::make("proxysrc", None)
+            let flow_combiner = FlowCombiner::new();
+
+            let audio_proxysrc = gst::ElementFactory::make("proxysrc", None)
+                .expect("Could not create proxysrc element");
+
+            let pad_templ = klass.get_pad_template("src").unwrap();
+            let audio_srcpad =
+                gst::GhostPad::new_no_target_from_template("audio_stream_src", &pad_templ).unwrap();
+            flow_combiner.add_pad(&audio_srcpad);
+
+            let video_proxysrc = gst::ElementFactory::make("proxysrc", None)
                 .expect("Could not create proxysrc element");
 
             let pad_templ = klass.get_pad_template("src").unwrap();
-            let srcpad =
-                gst::GhostPad::new_no_target_from_template("stream_src", &pad_templ).unwrap();
+            let video_srcpad =
+                gst::GhostPad::new_no_target_from_template("video_stream_src", &pad_templ).unwrap();
+            flow_combiner.add_pad(&video_srcpad);
+
+            audio_srcpad.set_chain_function(|_, _, _| {
+                Ok(gst::FlowSuccess::Ok)
+            });
+
+            video_srcpad.set_chain_function(|pad, parent, buffer| {
+            //    let result = gst::ProxyPad::chain_default::<gst::GhostPad, ServoMediaStreamSrc, Option<ServoMediaStreamSrc>>(pad, Some(parent), buffer);
```

Comment:
@sdroege it's not clear to me how to use [chain_default](https://slomo.pages.freedesktop.org/rustdocs/gstreamer/gstreamer/struct.ProxyPad.html#method.chain_default) here, and I couldn't find any examples of it. Could you advice on which associated types should I use here? Thanks!

Reply:
Why doesn't it compile with just
```rust
let result = gst::ProxyPad::chain_default(pad.downcast_ref::<gst::ProxyPad>().unwrap(), parent, buffer);
```

Reply:
Thanks! That worked

Reply:
Ugh, actually, it does not work.

```
error[E0283]: type annotations required: cannot resolve `_: glib::IsA<gst::Object>`
  --> backends/gstreamer/media_stream_source.rs:86:30
   |
86 |                 let result = gst::ProxyPad::chain_default(pad.upcast_ref::<gst::ProxyPad>(), parent, buffer);
   |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: required by `gst::proxy_pad::<impl gst::ProxyPad>::chain_default`
```

Reply:
`downcast_ref()`, not `upcast_ref()` :) You copied my code incorrectly. Not sure if *that* is the problem though. If what I pasted does not work please create an issue here https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/issues/new

It's supposed to be used like that and if it can't then that's a problem

Reply:
Yeah, I tried `downcast_ref()` first, but it was throwing this build error:
```
error[E0277]: the trait bound `gst::ProxyPad: glib::IsA<gst::GhostPad>` is not satisfied
  --> backends/gstreamer/media_stream_source.rs:85:70
   |
85 |                 let chain_result = gst::ProxyPad::chain_default(&pad.downcast_ref::<gst::ProxyPad>(), parent, buffer);
   |                                                                      ^^^^^^^^^^^^ the trait `glib::IsA<gst::GhostPad>` is not implemented for `gst::ProxyPad`
   |
   = help: the following implementations were found:
             <gst::ProxyPad as glib::IsA<glib::Object>>
             <gst::ProxyPad as glib::IsA<gst::Object>>
             <gst::ProxyPad as glib::IsA<gst::Pad>>
             <gst::ProxyPad as glib::IsA<gst::ProxyPad>>
   = note: required because of the requirements on the impl of `glib::object::CanDowncast<gst::ProxyPad>` for `gst::GhostPad`
```

Reply:
Oh you have a ghostpad there already.

```rust
let chain_result = gst::ProxyPad::chain_default(pad, parent, buffer);
```

Does not work? It seems like everything already has the right type for you.

Reply:
No, it doesn't work.

```
error[E0283]: type annotations required: cannot resolve `_: glib::IsA<gst::Object>`
  --> backends/gstreamer/media_stream_source.rs:85:36
   |
85 |                 let chain_result = gst::ProxyPad::chain_default(pad, parent, buffer);
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: required by `gst::proxy_pad::<impl gst::ProxyPad>::chain_default`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0283`.
error: Could not compile `servo-media-gstreamer`.
```

Reply:
Ok, can you create an issue about this then I'll take a look :) Ideally with a small testcase, a patch to the `gst-plugin-tutorial/src/progressbin.rs` for example

Reply:
Maybe too generic code here :)

Reply:
```rust
let chain_result = gst::ProxyPad::chain_default(pad, parent.as_ref(), buffer);
```

is the workaround for now btw. See https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/issues/203 and https://github.com/gtk-rs/gir/issues/777 for the effort to solve it in a nicer way everywhere else where this is wrong too.